File: CLI/src/main/java/wordhierarchyanalyzer/collections/tree/Tree.java
package wordhierarchyanalyzer.collections.tree;

import java.util.ArrayList;
import java.util.List;
import java.util.Optional;

public class Tree {

    private final Node root;

    public Tree() {
        this.root = new Node("", null, false);
    }

    private Node findNode(Node node, String text) {
        Node left = null;
        Node right = null;

        if (node.text.equalsIgnoreCase(text)) {
            return node;
        }

        if (node.nodeLeft != null) {
            left = findNode(node.nodeLeft, text);
        }

        if (left == null) {
            if (node.nodeRight != null) {
                right = findNode(node.nodeRight, text);
            }

            return right;
        }

        return left;
    }

    public void addText(String text, String textParent, boolean isAnalyzable) {
        Node nodeParent = null;
        Node nodeChild = null;

        nodeParent = this.findNode(this.root, textParent);

        if (nodeParent == null) {
            nodeParent = this.root;
        }

        nodeChild = new Node(text, nodeParent, isAnalyzable);

        if (nodeParent.nodeLeft == null) {
            nodeParent.nodeLeft = nodeChild;
        } else {
            Node node = nodeParent.nodeLeft;

            while (node.nodeRight != null) {
                node = node.nodeRight;
            }

            node.nodeRight = nodeChild;
        }
    }

    public Optional<String> depthLogic(String text, int depth) {
        List<String> texts = new ArrayList<>();
        Node node = this.findNode(this.root, text);

        if (node == null || !node.isAnalyzable) {
            return Optional.empty();
        }

        while (node.nodeParent != null) {
            texts.addFirst(node.text);

            node = node.nodeParent;
        }

        return Optional.of(texts.get(depth - 1));
    }

}

File: frontend/src/tree/tree.ts
import { TreeNode } from './node';

export class Tree {
    
    private root: TreeNode;
    private text: string;

    constructor() {
        this.root = new TreeNode("", null, false);
        this.text = "";
        this.getTree = this.getTree.bind(this);
    }

    private findTreeNode(treeNode: TreeNode, text: string): TreeNode | null {
        let left: TreeNode | null = null;
        let right: TreeNode | null = null;

        if (treeNode.text.toLowerCase() === text.toLowerCase()) {
            return treeNode;
        }

        if (treeNode.treeNodeLeft !== null) {
            left = this.findTreeNode(treeNode.treeNodeLeft, text);
        }

        if (left === null) {
            if (treeNode.treeNodeRight !== null) {
                right = this.findTreeNode(treeNode.treeNodeRight, text);
            }

            return right;
        }

        return left;
    }

    public addText(text: string, textParent: string, isAnalyzable: boolean): void {
        let treeNodeParent: TreeNode | null = this.findTreeNode(this.root, textParent) || this.root;
        let treeNodeChild: TreeNode = new TreeNode(text, treeNodeParent, isAnalyzable);

        if (treeNodeParent.treeNodeLeft === null) {
            treeNodeParent.treeNodeLeft = treeNodeChild;
        } else {
            let treeNode: TreeNode = treeNodeParent.treeNodeLeft;

            while (treeNode.treeNodeRight !== null) {
                treeNode = treeNode.treeNodeRight;
            }
            
            treeNode.treeNodeRight = treeNodeChild;
        }
    }

    public depthLogic(text: string, depth: number): string | null {
        let texts: string[] = [];
        let treeNode: TreeNode | null = this.findTreeNode(this.root, text);

        if (treeNode === null || !treeNode.isAnalyzable) {
            return null;
        }

        while (treeNode.treeNodeParent !== null) {
            texts.unshift(treeNode.text);
            treeNode = treeNode.treeNodeParent;
        }

        return texts[depth - 1] || null;
    }

    private getTreeAux(treeNode: TreeNode, tab: number): void {
        for (let i: number = 0; i < tab; i++) {
            this.text += "\t";
        }

        this.text += treeNode.text + "\n";

        if (treeNode.treeNodeLeft !== null) {
            this.getTreeAux(treeNode.treeNodeLeft, tab + 1);
        }

        if (treeNode.treeNodeRight !== null) {
            this.getTreeAux(treeNode.treeNodeRight, tab);
        }
    }

    public getTree(): string {
        this.text = "";
        this.getTreeAux(this.root, 0);

        return this.text;
    }

}
